File: internal/coordinator/tableregistration.go
package coordinator

import (
	"context"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strings"
	"time"

	sq "github.com/Masterminds/squirrel"
	"go.uber.org/zap"

	"github.com/y-scope/metalog/internal/db"
	"github.com/y-scope/metalog/internal/metastore"
	"github.com/y-scope/metalog/internal/schema"
)

// Sentinel errors for admin operations.
var (
	// ErrColumnNotFound is returned when no ACTIVE column matches the request.
	ErrColumnNotFound = errors.New("column not found")
	// ErrInvalidColumnPrefix is returned when column_name doesn't start with dim_f or agg_f.
	ErrInvalidColumnPrefix = errors.New("column name must start with dim_f or agg_f")
	// ErrInvalidAlias is returned when the alias value fails validation.
	ErrInvalidAlias = errors.New("invalid alias format")
)

// maxAliasLength is the maximum length of an alias column value.
const maxAliasLength = 128

// aliasPattern is the allowed pattern for alias values: alphanumeric, underscores, dots, hyphens, slashes.
var aliasPattern = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_./-]*$`)

// TableRegistration handles registering new tables at runtime.
type TableRegistration struct {
	db                  *sql.DB
	isMariaDB           bool
	compressionOverride string
	log                 *zap.Logger
}

// NewTableRegistration creates a TableRegistration.
func NewTableRegistration(db *sql.DB, isMariaDB bool, compressionOverride string, log *zap.Logger) *TableRegistration {
	return &TableRegistration{db: db, isMariaDB: isMariaDB, compressionOverride: compressionOverride, log: log}
}

// RegisterTableOpts holds optional fields for RegisterTable.
// Nil pointers mean "don't update" (keep DB default or existing value).
type RegisterTableOpts struct {
	ConfigJSON *string // JSON blob to merge into the config
}

// RegisterTable provisions a table and applies config overrides.
// Returns (created bool, err error). created is true if the table was newly provisioned.
func (s *TableRegistration) RegisterTable(
	ctx context.Context,
	tableName, displayName string,
	opts RegisterTableOpts,
) (bool, error) {
	if err := db.ValidateSQLIdentifier(tableName); err != nil {
		return false, err
	}

	// Check if table already exists
	existsQuery, existsArgs, _ := sq.Select("COUNT(*) > 0").
		From(metastore.TableRegistry).
		Where(sq.Eq{"table_name": tableName}).
		ToSql()
	var exists bool
	if err := s.db.QueryRowContext(ctx, existsQuery, existsArgs...).Scan(&exists); err != nil {
		return false, err
	}

	// Provision table
	if err := schema.EnsureTable(ctx, s.db, tableName, s.isMariaDB, s.compressionOverride, s.log); err != nil {
		return false, err
	}

	// Update display name if provided
	if displayName != "" {
		updateQuery, updateArgs, _ := sq.Update(metastore.TableRegistry).
			Set("display_name", displayName).
			Where(sq.Eq{"table_name": tableName}).
			ToSql()
		if _, err := s.db.ExecContext(ctx, updateQuery, updateArgs...); err != nil {
			return false, err
		}
	}

	// Update table config blob if any config fields are explicitly set.
	if err := s.updateTableConfig(ctx, tableName, opts); err != nil {
		return false, fmt.Errorf("update table config: %w", err)
	}

	created := !exists
	s.log.Info("table registered",
		zap.String("table", tableName),
		zap.Bool("created", created),
	)
	return created, nil
}

// updateTableConfig performs a read-modify-write on the config blob.
// Only fields explicitly set in opts are overridden.
func (s *TableRegistration) updateTableConfig(ctx context.Context, tableName string, opts RegisterTableOpts) error {
	if opts.ConfigJSON == nil {
		return nil
	}

	// Read existing config blob.
	query, args, _ := sq.Select("config").
		From(metastore.TableRegistryConfig).
		Where(sq.Eq{"table_name": tableName}).
		ToSql()
	var blob []byte
	if err := s.db.QueryRowContext(ctx, query, args...).Scan(&blob); err != nil && err != sql.ErrNoRows {
		return err
	}

	cfg, err := metastore.DecodeTableConfig(blob)
	if err != nil {
		return err
	}

	// Merge JSON overrides into existing config.
	// DisallowUnknownFields catches typos like "consolidation_enbaled".
	dec := json.NewDecoder(strings.NewReader(*opts.ConfigJSON))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&cfg); err != nil {
		return fmt.Errorf("parse config_json: %w", err)
	}

	// Encode and write back.
	newBlob, err := metastore.EncodeTableConfig(cfg)
	if err != nil {
		return err
	}

	updateQuery, updateArgs, _ := sq.Update(metastore.TableRegistryConfig).
		Set("config", newBlob).
		Where(sq.Eq{"table_name": tableName}).
		ToSql()
	_, err = s.db.ExecContext(ctx, updateQuery, updateArgs...)
	return err
}

// resolveRegistryTable returns the registry table name and key column for a
// given column name based on its prefix (dim_f or agg_f).
func resolveRegistryTable(colName string) (registryTable, keyColumn string, err error) {
	if strings.HasPrefix(colName, metastore.DimColumnPrefix) {
		return metastore.DimRegistryTable, "dim_key", nil
	}
	if strings.HasPrefix(colName, metastore.AggColumnPrefix) {
		return metastore.AggRegistryTable, "agg_key", nil
	}
	return "", "", fmt.Errorf("%w: %q (must start with %q or %q)",
		ErrInvalidColumnPrefix, colName, metastore.DimColumnPrefix, metastore.AggColumnPrefix)
}

// validateAlias checks alias format. Returns the trimmed alias or an error.
func validateAlias(alias string) (string, error) {
	alias = strings.TrimSpace(alias)
	if alias == "" {
		return "", nil // empty alias clears the value
	}
	if len(alias) > maxAliasLength {
		return "", fmt.Errorf("%w: exceeds max length of %d characters", ErrInvalidAlias, maxAliasLength)
	}
	if !aliasPattern.MatchString(alias) {
		return "", fmt.Errorf("%w: must match [a-zA-Z_][a-zA-Z0-9_./-]*", ErrInvalidAlias)
	}
	return alias, nil
}

// SetColumnAlias sets or clears the alias for a dimension or aggregation column.
// Returns the validated alias. Returns ErrColumnNotFound if no ACTIVE column matches.
func (s *TableRegistration) SetColumnAlias(ctx context.Context, tableName, colName, alias string) (string, error) {
	alias, err := validateAlias(alias)
	if err != nil {
		return "", err
	}

	registryTable, _, err := resolveRegistryTable(colName)
	if err != nil {
		return "", err
	}

	// Atomic update — only touches ACTIVE rows.
	var aliasVal any
	if alias != "" {
		aliasVal = alias
	}
	updateQuery, updateArgs, _ := sq.Update(registryTable).
		Set("alias_column", aliasVal).
		Where(sq.Eq{"table_name": tableName, "column_name": colName, "state": "ACTIVE"}).
		ToSql()
	res, err := s.db.ExecContext(ctx, updateQuery, updateArgs...)
	if err != nil {
		return "", fmt.Errorf("update alias: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return "", fmt.Errorf("rows affected: %w", err)
	}
	if affected == 0 {
		return "", fmt.Errorf("%w: no ACTIVE column %s in table %s", ErrColumnNotFound, colName, tableName)
	}

	s.log.Info("column alias updated",
		zap.String("table", tableName),
		zap.String("column", colName),
		zap.String("alias", alias),
	)

	return alias, nil
}

// InvalidateColumn marks a dimension or aggregation column as INVALIDATED.
// Returns the previous key of the invalidated column.
func (s *TableRegistration) InvalidateColumn(ctx context.Context, tableName, colName string) (string, error) {
	registryTable, keyColumn, err := resolveRegistryTable(colName)
	if err != nil {
		return "", err
	}

	// Read the current key before invalidating.
	var previousKey string
	selectQuery, selectArgs, _ := sq.Select(keyColumn).
		From(registryTable).
		Where(sq.Eq{"table_name": tableName, "column_name": colName, "state": "ACTIVE"}).
		ToSql()
	if err := s.db.QueryRowContext(ctx, selectQuery, selectArgs...).Scan(&previousKey); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return "", fmt.Errorf("%w: no ACTIVE column %s in table %s", ErrColumnNotFound, colName, tableName)
		}
		return "", fmt.Errorf("lookup column: %w", err)
	}

	// Transition ACTIVE → INVALIDATED.
	now := time.Now().UnixNano()
	updateQuery, updateArgs, _ := sq.Update(registryTable).
		Set("state", "INVALIDATED").
		Set("invalidated_at", now).
		Where(sq.Eq{"table_name": tableName, "column_name": colName, "state": "ACTIVE"}).
		ToSql()
	res, err := s.db.ExecContext(ctx, updateQuery, updateArgs...)
	if err != nil {
		return "", fmt.Errorf("invalidate column: %w", err)
	}
	affected, err := res.RowsAffected()
	if err != nil {
		return "", fmt.Errorf("rows affected: %w", err)
	}
	if affected == 0 {
		return "", fmt.Errorf("%w: column %s in table %s was invalidated concurrently",
			ErrColumnNotFound, colName, tableName)
	}

	s.log.Info("column invalidated",
		zap.String("table", tableName),
		zap.String("column", colName),
		zap.String("previousKey", previousKey),
	)

	return previousKey, nil
}

File: crates/metalog-coordinator/src/table_registration.rs
use metalog_schema::ensure_table;
use sqlx::MySqlPool;

/// Handles runtime table provisioning and configuration.
pub struct TableRegistration {
    db: MySqlPool,
    _compression_override: String,
}

/// Options for registering a table.
#[derive(Debug, Default)]
pub struct RegisterTableOpts {
    /// JSON config to merge into the stored config blob.
    pub config_json: Option<String>,
}

impl TableRegistration {
    pub fn new(db: MySqlPool, compression_override: &str) -> Self {
        Self {
            db,
            _compression_override: compression_override.to_string(),
        }
    }

    /// Registers a new table. Returns true if newly created, false if already exists.
    ///
    /// Idempotent: calling with the same table name multiple times is safe.
    pub async fn register_table(
        &self,
        table_name: &str,
        _display_name: &str,
        opts: RegisterTableOpts,
    ) -> Result<bool, TableRegistrationError> {
        let created = ensure_table(&self.db, table_name, opts.config_json.as_deref())
            .await
            .map_err(TableRegistrationError::Provision)?;

        Ok(created)
    }

    /// Sets or clears the alias for a dimension or aggregation column.
    pub async fn set_column_alias(
        &self,
        table_name: &str,
        column_name: &str,
        alias: &str,
    ) -> Result<String, TableRegistrationError> {
        let alias_val = if alias.is_empty() {
            None
        } else {
            validate_alias(alias)?;
            Some(alias)
        };

        let registry_table = resolve_registry_table(column_name)?;

        let sql = format!(
            "UPDATE {registry_table} SET alias_column = ? WHERE table_name = ? AND column_name = \
             ? AND state = 'ACTIVE'"
        );
        let result = sqlx::query(&sql)
            .bind(alias_val)
            .bind(table_name)
            .bind(column_name)
            .execute(&self.db)
            .await
            .map_err(TableRegistrationError::Sql)?;

        if result.rows_affected() == 0 {
            return Err(TableRegistrationError::ColumnNotFound(
                column_name.to_string(),
            ));
        }

        Ok(alias_val.unwrap_or("").to_string())
    }

    /// Marks a column as INVALIDATED.
    pub async fn invalidate_column(
        &self,
        table_name: &str,
        column_name: &str,
    ) -> Result<String, TableRegistrationError> {
        let registry_table = resolve_registry_table(column_name)?;
        let now = metalog_timeutil::epoch_nanos();

        // Get the current key before invalidating.
        let key_col = if column_name.starts_with("dim_") {
            "dim_key"
        } else {
            "agg_key"
        };
        let select_sql = format!(
            "SELECT {key_col} FROM {registry_table} WHERE table_name = ? AND column_name = ? AND \
             state = 'ACTIVE'"
        );
        let key: Option<(String,)> = sqlx::query_as(&select_sql)
            .bind(table_name)
            .bind(column_name)
            .fetch_optional(&self.db)
            .await
            .map_err(TableRegistrationError::Sql)?;

        let previous_key = key
            .map(|r| r.0)
            .ok_or_else(|| TableRegistrationError::ColumnNotFound(column_name.to_string()))?;

        let update_sql = format!(
            "UPDATE {registry_table} SET state = 'INVALIDATED', invalidated_at = ? WHERE \
             table_name = ? AND column_name = ? AND state = 'ACTIVE'"
        );
        sqlx::query(&update_sql)
            .bind(now)
            .bind(table_name)
            .bind(column_name)
            .execute(&self.db)
            .await
            .map_err(TableRegistrationError::Sql)?;

        tracing::info!(
            table = table_name,
            column = column_name,
            previous_key = %previous_key,
            "invalidated column"
        );

        Ok(previous_key)
    }
}

fn resolve_registry_table(column_name: &str) -> Result<&'static str, TableRegistrationError> {
    if column_name.starts_with("dim_") {
        Ok("_dim_registry")
    } else if column_name.starts_with("agg_") {
        Ok("_agg_registry")
    } else {
        Err(TableRegistrationError::InvalidColumnPrefix(
            column_name.to_string(),
        ))
    }
}

fn validate_alias(alias: &str) -> Result<(), TableRegistrationError> {
    if alias.len() > 128 {
        return Err(TableRegistrationError::InvalidAlias(
            "alias too long (max 128 chars)".into(),
        ));
    }
    // Allow alphanumeric, underscore, dot, hyphen, slash.
    if !alias
        .chars()
        .all(|c| c.is_alphanumeric() || c == '_' || c == '.' || c == '-' || c == '/')
    {
        return Err(TableRegistrationError::InvalidAlias(format!(
            "alias contains invalid characters: {alias}"
        )));
    }
    Ok(())
}

#[derive(Debug, thiserror::Error)]
pub enum TableRegistrationError {
    #[error("provision: {0}")]
    Provision(#[source] metalog_schema::EnsureTableError),

    #[error("sql: {0}")]
    Sql(#[source] sqlx::Error),

    #[error("column not found: {0}")]
    ColumnNotFound(String),

    #[error("invalid column prefix: {0} (expected dim_ or agg_)")]
    InvalidColumnPrefix(String),

    #[error("invalid alias: {0}")]
    InvalidAlias(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_dim_registry() {
        assert_eq!(resolve_registry_table("dim_f01").unwrap(), "_dim_registry");
    }

    #[test]
    fn resolve_agg_registry() {
        assert_eq!(resolve_registry_table("agg_f01").unwrap(), "_agg_registry");
    }

    #[test]
    fn resolve_invalid_prefix() {
        assert!(resolve_registry_table("invalid_col").is_err());
    }

    #[test]
    fn validate_alias_valid() {
        assert!(validate_alias("my_alias").is_ok());
        assert!(validate_alias("path/to/field").is_ok());
        assert!(validate_alias("field-name").is_ok());
        assert!(validate_alias("field.name").is_ok());
    }

    #[test]
    fn validate_alias_invalid() {
        assert!(validate_alias("has space").is_err());
        assert!(validate_alias(&"a".repeat(129)).is_err());
    }

    #[test]
    fn error_display() {
        let err = TableRegistrationError::ColumnNotFound("dim_f99".into());
        assert!(err.to_string().contains("dim_f99"));
    }
}
